Add shuffle Spark function

zhli1142015 opened the pull request:
Doc: https://spark.apache.org/docs/latest/api/sql/#shuffle
Code: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L1257

Spark `shuffle` uses a specific seed for random function, which is the sum 
of the `seed` provided by the user and the configuration `spark.partition_id`, 
this is the different from presto's behavior. So we need to refactor existing 
`shuffle` to allow it to accept random / custom seed.

zhli1142015 commented:
cc @PHILO-HE  and @rui-mo , thanks.

rui-mo commented on velox/functions/sparksql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    const auto seed = inputArgs[1]
+                          .constantValue->template as<ConstantVector<int64_t>>()
+                          ->valueAt(0);
```
nit: do we need to verify arg 1 is constant before the access?

zhli1142015 replied:
Added, thanks.

rui-mo commented on velox/functions/sparksql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,66 @@
+    const auto seed = inputArgs[1]
+                          .constantValue->template as<ConstantVector<int64_t>>()
+                          ->valueAt(0);
+    const int32_t partitionId = config.sparkPartitionId();
```
nit: could we use `seed + config.sparkPartitionId()` instead and remove this variable?

PHILO-HE reviewed (commented):
Looks good! cc @rui-mo

rui-mo reviewed (commented):
Thanks.

zhli1142015 commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,140 @@
+
+using namespace facebook::velox::test;
+
+using facebook::velox::functions::test::FunctionBaseTest;
```
Removed, thanks.

zhli1142015 commented on velox/functions/sparksql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,66 @@
+      const std::vector<exec::VectorFunctionArg>& inputArgs,
+      const core::QueryConfig& config) {
+    VELOX_CHECK_GE(inputArgs.size(), 2);
+    VELOX_CHECK_EQ(inputArgs[1].type->kind(), TypeKind::BIGINT);
```
Sure, thanks.

rui-mo reviewed (commented):
Thanks. Looks good to me.

zhli1142015 commented:
@mbasmanova , could you help to review this PR? Thanks.

jinchengchenghh commented:
What's the difference with before? Why not just register the function in Spark? This PR moves too much code to header file.

zhli1142015 commented:
> What's the difference with before? Why not just register the function in Spark? This PR moves too much code to header file.
Spark `array_shuffle`  uses a specific seed for random function, which is the sum of the `seed` provided by the user and the configuration `spark.partition_id`, this is the different from presto's behavior.

jinchengchenghh commented:
> > What's the difference with before? Why not just register the function in Spark? This PR moves too much code to header file.
> 
> Spark `array_shuffle` uses a specific seed for random function, which is the sum of the `seed` provided by the user and the configuration `spark.partition_id`, this is the different from presto's behavior.

Can you add the explanation to PR description?

zhli1142015 commented:
> > > What's the difference with before? Why not just register the function in Spark? This PR moves too much code to header file.
> > 
> > 
> > Spark `array_shuffle` uses a specific seed for random function, which is the sum of the `seed` provided by the user and the configuration `spark.partition_id`, this is the different from presto's behavior.
> 
> Can you add the explanation to PR description?

Sure, updated,

jinchengchenghh commented:
You could register it in `workAroundRegistrationMacro`, still don't know why move to header file

zhli1142015 commented:
> You could register it in `workAroundRegistrationMacro`, still don't know why move to header file

The original function couldn't accept a random seed parameter, so we modified it to allow for specified random seed parameters. Since this function is used by both Presto and Spark, we moved it to the lib folder. I don't notice any issue about the  `workAroundRegistrationMacro`, is there some thing I need to change?

jinchengchenghh commented:
> > You could register it in `workAroundRegistrationMacro`, still don't know why move to header file
> 
> The original function couldn't accept a random seed parameter, so we modified it to allow for specified random seed parameters. Since this function is used by both Presto and Spark, we moved it to the lib folder. I don't notice any issue when not reigister it in the `workAroundRegistrationMacro`, is there some thing I need to change?

Spark folder can also invoke Prestosql implement, you can refer to https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/Register.cpp#L88

The implement locates in prestosql but both used in spark and presto
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/FilterFunctions.cpp#L242

zhli1142015 commented:
> > > You could register it in `workAroundRegistrationMacro`, still don't know why move to header file
> > 
> > 
> > The original function couldn't accept a random seed parameter, so we modified it to allow for specified random seed parameters. Since this function is used by both Presto and Spark, we moved it to the lib folder. I don't notice any issue when not reigister it in the `workAroundRegistrationMacro`, is there some thing I need to change?
> 
> Spark folder can also invoke Prestosql implement, you can refer to https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/Register.cpp#L88
> 
> The implement locates in prestosql but both used in spark and presto https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/FilterFunctions.cpp#L242

Thanks, I revisit other shared functions' implementation. I think I should follow function `repeat`, as spark and presto function signature are different.

jinchengchenghh commented on velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:
```diff
@@ -135,7 +136,11 @@ void registerArrayFunctions(const std::string& prefix) {
   VELOX_REGISTER_VECTOR_FUNCTION(udf_zip_with, prefix + "zip_with");
   VELOX_REGISTER_VECTOR_FUNCTION(udf_array_position, prefix + "array_position");
-  VELOX_REGISTER_VECTOR_FUNCTION(udf_array_shuffle, prefix + "shuffle");
+  exec::registerStatefulVectorFunction(
```
It is stateless

zhli1142015 replied:
Thanks, it should be stateful after this change, the state should be `randGen_`, we want it be assigned a new  random seed  for every expr compilation.

jinchengchenghh commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,138 @@
+};
+
+TEST_F(ArrayShuffleTest, basic) {
+  auto input = makeNullableArrayVector<int64_t>({{1, 2, 3, 4, 5}});
```
makeNullableArrayVector not nullable

jinchengchenghh commented on velox/docs/functions/spark/array.rst:
```diff
@@ -86,6 +86,16 @@ Array Functions
+    from the user-specified ``seed`` and the configuration `spark.partition_id`.
+    ``seed`` must be constant. ::
+
+        SELECT array_shuffle(ARRAY [1, 2, 3], 0); -- [3, 1, 2]
```
This is not a valid SQL in Spark. 
```
spark-sql (default)> SELECT shuffle(ARRAY (1, 2, 3), 0);
[WRONG_NUM_ARGS.WITHOUT_SUGGESTION] The `shuffle` requires 1 parameters but the actual number is 2. Please, refer to 'https://spark.apache.org/docs/latest/sql-ref-functions.html' for a fix.; line 1 pos 7
spark-sql (default)> SELECT shuffle(ARRAY (1, 2, 3));
[3,2,1]
Time taken: 0.193 seconds, Fetched 1 row(s)
spark-sql (default)> SELECT shuffle(ARRAY (1, 2, 3,5));
[3,1,2,5]
Time taken: 0.139 seconds, Fetched 1 row(s)
```
We could change to
`
SELECT array_shuffle(array(1, 2, 3), 0); -- [3, 1, 2]`

zhli1142015 replied:
Yes, this is not a valid SQL in spark, spark generated the seed and pass to this function, use can't set the seed.

jinchengchenghh replied:
ARRAY [1, 2, 3]  -> array(1, 2, 3)

jinchengchenghh commented on velox/functions/lib/ArrayShuffle.cpp:
```diff
@@ -112,12 +117,41 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  return std::make_unique<ArrayShuffleFunction>(std::random_device{}());
+}
+
+std::shared_ptr<exec::VectorFunction> makeArrayShuffleWithCustomSeed(
```
Does this function should move to `sparkql` directory? If that's the case, other similar function should obey this rule.

jinchengchenghh replied:
hi @mbasmanova Which way do you think is better?

mbasmanova replied:
This pattern seems fine to me.

jinchengchenghh commented:
Please highlight the function is `shuffle` in Spark but `array_shuffle` to add `seed` argument in PR description. By the way, I suppose we could use `shuffle` instead.

zhli1142015 commented:
@jinchengchenghh  and @mbasmanova , do you have more comments for this PR?
Thanks.

jinchengchenghh commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,136 @@
+  outerArrayType row3{{}};
+  outerArrayType row4{{{std::nullopt}}};
+  auto input =
+      makeNullableNestedArrayVector<int64_t>({{row1}, {row2}, {row3}, {row4}});
```
same, not nullable

zhli1142015 replied:
Changed to `makeNestedArrayVectorFromJson`, thanks.

jinchengchenghh commented:
LGTM Thanks!

zhli1142015 commented:
@mbasmanova  and @Yuhta , could you help review this PR? Thanks.
BTW, looks CI failure is not related to this PR.

mbasmanova commented:
> looks CI failure is not related to this PR.

Is there an existing issue about this failure? If not, would you create one? 

CC: @kgpai

assignUser commented:
The CI failure is due to an issue with the step generating the signatures on main. I'll open a PR.

zhli1142015 commented:
CI jobs pass, @mbasmanova  could you please help check this PR?
Thanks.

mbasmanova commented on velox/functions/lib/ArrayShuffle.cpp:
```diff
@@ -112,12 +117,41 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  const auto seed =
+      inputArgs[1]
+          .constantValue->template as<ConstantVector<int64_t>>()
+          ->valueAt(0);
```
Need to check for null as well.

zhli1142015 replied:
Added, thanks.

mbasmanova commented on velox/functions/lib/ArrayShuffle.cpp:
```diff
@@ -112,12 +117,41 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+
+exec::VectorFunctionMetadata arrayShuffleMetadata() {
+  return exec::VectorFunctionMetadataBuilder()
+      .deterministic(false)
```
It seems like Spark's version is deterministic, no?

zhli1142015 replied:
Updated, thanks.

zhli1142015 replied:
Sorry my bad, I think it should be marked as non-deterministic to prevent constant folding. If it's deterministic, the output for `wrapInConstant(3, [1, 2, 3])` would be `{[3, 2, 1], [3, 2, 1], [3, 2, 1]}`, this is not expected. Please let me know if this makes sense to you.

mbasmanova replied:
@zhli1142015 I see. Makes sense.

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,129 @@
+  auto result = makeArrayVector<int64_t>({{3, 5, 4, 1, 2}});
+  auto stringInput = makeArrayVector<std::string>({{"a", "b", "c", "d"}});
+  auto stringResult = makeArrayVector<std::string>({{"a", "c", "b", "d"}});
+  compareResult<int64_t>(testShuffle<int64_t>(input, 0), result, true);
```
Since this function is deterministic, we can simply hard-code the results and use assertEqualVectors helper function.

zhli1142015 replied:
Updated, thanks.

mbasmanova reviewed (commented):
@zhli1142015 Looks good overall. Some minor comments.

zhli1142015 commented:
Run Spark fuzzer for 30 mins. Thanks.
```
E0508 19:45:11.026609 396279 ExpressionFuzzerVerifier.cpp:373] Total iterations: 9654004
E0508 19:45:11.026638 396279 ExpressionFuzzerVerifier.cpp:374] Total failed: 524124
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,117 @@
+  assertEqualVectors(testShuffle<std::string>(stringInput, 0), stringResult);
+
+  // Assert results are different with different seeds / partition ids.
+  assertDifferentSequence<int64_t>(
```
Since this function is deterministic we can simply hard-code expected results. This would make the test easier to read and also make it stronger. The test would fail if function is modified in a way that makes it non-deterministic.

mbasmanova commented on velox/functions/lib/ArrayShuffle.h:
```diff
@@ -0,0 +1,50 @@
+namespace facebook::velox::functions {
+
+std::vector<std::shared_ptr<exec::FunctionSignature>>
+arrayShuffleWithSignatures();
```
drop 'With': arrayShuffleSignatures

mbasmanova commented on velox/functions/lib/ArrayShuffle.h:
```diff
@@ -0,0 +1,50 @@
+// with 'makeArrayShuffleWithCustomSeed'.
+// This function needs to generate different sequences for the constant input,
+// set 'deterministic' as false to prevent the constant folding.
+exec::VectorFunctionMetadata getMetadataForArrayShuffleWithCustomSeed();
```
Since both functions return the same values, let's consolidate into one: getMetadataForArrayShuffle

mbasmanova commented on velox/functions/lib/ArrayShuffle.h:
```diff
@@ -0,0 +1,50 @@
+std::vector<std::shared_ptr<exec::FunctionSignature>>
+arrayShuffleWithSignatures();
+
+// This function returns metadata with 'deterministic' as false, it is used
```
Drop 'This function returns'. Use /// for comments on public APIs

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  template <typename T>
+  void assertDifferentSequence(
```
This method is not needed. Let's use assertEqualVectors.

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,110 @@
+
+class ArrayShuffleTest : public SparkFunctionBaseTest {
+ protected:
+  template <typename T>
```
This template parameter is not used. Let's remove.

Alternatively, we may change input to std::string that contains array value as JSON. This would simply call sites a bit.

```
assertEqualVectors(testShuffle<int64_t>("[1, 2, 3, 4, 5]", 0), result);
```

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,110 @@
+TEST_F(ArrayShuffleTest, basic) {
+  auto input = makeArrayVector<int64_t>({{1, 2, 3, 4, 5}});
+  auto result = makeArrayVector<int64_t>({{3, 5, 4, 1, 2}});
+  auto resultForPartitionIdOne = makeArrayVector<int64_t>({{2, 1, 3, 4, 5}});
```
The names of resultForPartitionIdOne and resultForSeesTwo are long and a bit hard to reason about. Consider removing them altogether.

```
assertEqualVectors(
      testShuffle<int64_t>(input, 0, 1), makeArrayVector<int64_t>({{2, 1, 3, 4, 5}}));

assertEqualVectors(
      testShuffle<int64_t>(input, 2, 0), makeArrayVector<int64_t>({{4, 1, 3, 5, 2}}));
```

Alternatively, we can enhance testShuffle to take 2 JSON strings for input and result arrays:

```
testShuffle<int64_t>("[1, 2, 3, 4, 5]", 0, 1), "[2, 1, 3, 4, 5]");
```

zhli1142015 replied:
Thanks.

mbasmanova reviewed (commented):
@zhli1142015 Looks good. Some suggestions for the test.

mbasmanova commented on velox/functions/sparksql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,125 @@
+TEST_F(ArrayShuffleTest, basic) {
+  testShuffle<int64_t>("[1, 2, 3, 4, 5]", "[3, 5, 4, 1, 2]", 0);
+  testShuffle<std::string>(
+      "[\"a\", \"b\", \"c\", \"d\"]", "[\"a\", \"c\", \"b\", \"d\"]", 0);
```
use R"(["a", "b", "c", "d"])" for readability

zhli1142015 replied:
Thanks.

mbasmanova reviewed (approved):
Thanks.
